### PR TITLE
Allow defining customized manual and/or automatic link decorators

### LINKED ISSUE #186: Allow setting link's target attribute (allow opening links in new tabs)
## Original ticket description

> Hello!
>
> There are not ability to select target of link in ckeditor5 . 
>
> Please, add it

## Feedback needed

Please read this discussion starting from [this comment](https://github.com/ckeditor/ckeditor5-link/issues/186#issuecomment-452392660) and leave your feedback (by adding a reaction) in [this comment](https://github.com/ckeditor/ckeditor5-link/issues/186#issuecomment-452608537).

### PR DESCRIPTION
**Pull request by msamsel**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Allow on defining customized manual and/or automatic link decorators. Closes #186 .

---

### Additional information

Required theme-lark PR [t/ckeditor5-link/186](https://github.com/ckeditor/ckeditor5-theme-lark/pull/233)

* Currently you cannot use both option at the same time when those works on the same attribute.

### CODE REVIEW

**Comment by msamsel** [comment]
There still remain documentation guide to write.
And currently it's pretty simple UI, maybe we could invent some better approach for that.
Screen Shot 2019-04-30 at 14 03 24

**Comment by mlewand** [comment]
@msamsel what is the reason for graying out 2 options? Is it that these are the options set up automatically? If so I think it is safe to skip them in the UI and make them only happen _behind the scenes_.

@dkonopka could you have a look at this from UI/UX point of view? Though the components are aligned with our design in terms of styles I feel that this proposal makes the dialog very heavy. IMO it doesn't look nice if there are more than 2 options. Part of which might be displaying these grayed out options.

**Comment by msamsel** [comment]
> @msamsel what is the reason for graying out 2 options? Is it that these are the options set up automatically? If so I think it is safe to skip them in the UI and make them only happen behind the scenes.

I haven't configure it and it is just default behavior. As this is `SwitchButton` instance its just indicate that this button is turned on. And gets highlight like bold button or any other regular button.

But it might be somehow improved I see that options for table are not highlighted, so there should be some way to not showing it up.
Screen Shot 2019-05-14 at 12 11 38

**Comment by mlewand** [comment]
So definitely we need to unify that, my first impression was that the last two options were simply disabled. Whereas as I understood from your previous comment it means that the first option is hovered (thus it's white) and others are in their default state (gray background), right?

**Comment by msamsel** [comment]
@mlewand unfortunately couldn't get mouse pointer on this screenshot. Options have this state:
1. has no interaction
2. is active (turned on)
3. is hovered

There is manual test, which has an example of manual decorators. You should be able to get live demo with this menu if you need check more details.

**Comment by msamsel** [comment]
@mlewand it's appear that feature that switchbutton doesn't have background inside list is available in CKE5:
https://github.com/ckeditor/ckeditor5-theme-lark/blob/d3e4266526b277a7c1e73da91f8e5c8ddf38e7df/theme/ckeditor5-ui/components/list/list.css#L64-L74

So I just give proper classes to my component and import additional style to manual test, and everything is working nicely without additional hacks.

**Comment by msamsel** [comment]
@mlewand now grey background is not present for active switch buttons. I reuse class names already available in our theme.

**Comment by mlewand** [comment]
The UX of ths feature needs to be polished.

* Switches should apply the change the moment user clicks it, no clicking accept button should be required.
* I think that the toggle buttons provided by this feature should also be visible on a first link view (where you have clickable link). I find it really more difficult that it should be to get the job done when options are hidden in second view.

As a reference, a gif from the current implementation:

image

## RFC

Should links be applied to the "link preview" visible on the first view? I'm thinking mostly about the download attribute here. It would be nice if the link gets the `[download]` attribute. Another useful attribute might be [rel="noreferrer"].

But at the same time I find it useful only for these particular attributes. Other, like HTML class doesn't make much sense in this regard.

**Comment by msamsel** [comment]
> Should links be applied to the "link preview" visible on the first view?

I have one doubt about it. When someone creates new link, by default it's opened form view. So in this case it will extend his path. He will have to create new link and then close edition to go to action menu and apply proper decorators. It sounds wrong for me.

Another case is that those menus are grouped by function: "action" and "edititng". Providing new decorator to link is editing from my perspective.

**Comment by poppaul95** [comment]
@msamsel can you provide some eta for when this feature will be available

**Comment by dkonopka** [comment]
> I think that the toggle buttons provided by this feature should also be visible on a first link view (where you have a clickable link)

I know it's handy to have all need things at the first view, but I can't agree with you.  

It's a 2-step form (preview & editing), so we should teach users if they want to change anything they should go to the "edit" section.

OTOH, I'm not so sure about "switch" buttons - it looks good as a single option (for tables).

I'm wondering, maybe we should consider a new UI element - a checkbox?  We definitely need an opinion from @oleq here, but at this moment we can live with those switches?
 
decorators

**Comment by msamsel** [comment]
@poppaul95 I'm sorry there's no precise ETA yet. Initially we were looking to include it in iteration 24th, however we already know that we'll have to put it into iteration 25.

The best way to stay up to date is to subscribe ckeditor/ckeditor5-link#186 issue as we'll update it as the feature lands in the `master` branch.

**Comment by mlewand** [comment]
Interesting point @dkonopka. I think that in this case we should wait to hear more voices on the design.

I'm still not convinced whether we should put these toggles in the main view (or both) or in the second.

Talking about two view approach, I see couple of problems with this idea:

* Accept/cancel buttons position
* As a user I'm not feeling comfortable with a thought of changing target attribute in more than one link in my document.
* Closing confirmation

> OTOH, I'm not so sure about "switch" buttons - it looks good as a single option (for tables).

This should not be our concern at this point, we're giving this option to the developers to see whether this feature will be used.

## Accept/cancel buttons position

With new inputs added below, I believe that the ok/cancel buttons should be moved to the very bottom of the balloon. I mean we could go unusual here, but it's a common solution to put action buttons at the bottom of the dropdown if there are multiple options in it.

## Closing confirmation

My initial impression was that attribute gets applied the moment I toggle it. So I just did that and moved on. Later I was surprised that my changes were not applied.

For that reasons apps typically put some sort of confirmation when you're closing "dirty" dropdown/dialog. This would have tipped me of this behavior.

**Comment by msamsel** [comment]
> My initial impression was that attribute gets applied the moment I toggle it. So I just did that and moved on. Later I was surprised that my changes were not applied.

That's true. Currently entire view works as a form, you need to submit changes with ok button, to apply changes. It's made to be perfectly integrate with "link" command, as it was one of point discussed here:
https://github.com/ckeditor/ckeditor5-link/issues/186#issuecomment-483257606

So now all link attributes (href and custom decorators) are passed inside link command.

**Comment by msamsel** [comment]
I've rebased this PR to the current master to fix an appearing conflicts.

File: src/ui/linkformview.js [hunk 1]
```diff
@@ -210,6 +240,79 @@ export default class LinkFormView extends View {
...
+	_createCustomAttributesView() {
+		const switches = this.createCollection();
+		const t = this.locale.t;
+
+		switches.bindTo( this.customAttributes ).using( item => {
+			const switchButton = new SwitchButtonView( this.locale );
+
+			switchButton.set( {
+				name: item.id,
+				label: t( item.label ),
```
Comment (msamsel) [inline src/ui/linkformview.js]: It's a wrong `t()` usage it has to be done with a string.

**Comment by oleq** [comment]
**Disclaimer**: I didn't check the code, just the UI/UX.

I found some obvious rendering issues in this PR like this one:

Screenshot 2019-06-11 at 14 13 47

Or this one. Please keep in mind that we support RWD in certain balloons/forms. 

image

But they don't matter so much because there's plenty of work in terms of general UI/UX here anyway:

- You can't have your cake and eat it too. Our compact form design with a horizontal layout (`input | 2 buttons`) works as long as there's nothing more to be displayed there. When additional form elements show up, there must be a change in the layout.
- We already support a vertical layout of the form in the RWD and it's time to re-use it in a situation when decorators kick in. 
   image
- It's OK to use switch toggles.
- We can bring back the labels to "Save" and "Cancel" buttons since the form is already quite wide.
- The list of decorators should be a bit denser than in the PR to save vertical space.
- Long story short, the design should be more or less like this
   image
- ...or, in a more realistic use–case (because it's mostly about `target="..."`, let's be honest):
   image
- ☝️ It means that the decorator container should be placed before the Save/Cancel buttons in the form's `ViewCollection`

File: src/ui/linkformview.js [hunk 2]
```diff
@@ -77,6 +84,31 @@ export default class LinkFormView extends View {
 		 */
 		this.cancelButtonView = this._createButton( t( 'Cancel' ), cancelIcon, 'ck-button-cancel', 'cancel' );
 
+		/**
+		 * Keeps reference to {@link module:link/linkcommand~LinkCommand#customAttributes}.
+		 *
+		 * @readonly
+		 * @type {model:utils/collection~Collection}
```
Comment (oleq) [inline src/ui/linkformview.js]: Collection of what?

**Comment by dkonopka** [comment]
> We already support a vertical layout of the form in the RWD and it's time to re-use it in a situation when decorators kick in.

👍 

> The list of decorators should be a bit denser than in the PR to save vertical space.

👍 

> We can bring back the labels to "Save" and "Cancel" buttons since the form is already quite wide.

👍 It looks nice in the EN example, but I'm a bit worried about localization, how it looks for other languages, but if the form will be stretching then it's ok.

**To be sure**: we are not changing the default design of link form when there are no "link decorators"?

**Comment by oleq** [comment]
> To be sure: we are not changing the default design of link form when there are no "link decorators"?

Yes. We switch from h- to v-layout when decorators show up.

File: src/linkcommand.js [hunk 3]
```diff
@@ -33,6 +48,11 @@ export default class LinkCommand extends Command {
 		const doc = model.document;
 
 		this.value = doc.selection.getAttribute( 'linkHref' );
+
+		for ( const customAttr of this.customAttributes ) {
+			customAttr.value = doc.selection.getAttribute( customAttr.id ) || false;
```
[the remark this answers is not in the document]
Comment (msamsel) [reply src/linkcommand.js]: > Why are all the manual decorators updated upon each LinkCommand#refresh()? Is this required by some logic?

To synchronize SwitchButton when you move selection between links. It is a similar situation to updating value of link href in balloon panel.

> Could their values be obtained on demand, for instance as a getter?

We need to update state of all decorator to propagate current state to view so we need to check all of them anyway.

Do you think is better to implement some logic with getters for this kind of situation?

**Comment by dkonopka** [comment]
I've updated the link form view with manual decorators according to your suggestions @oleq. 

One thing I've skipped is the text label next to the check and cross icons because I'm worried about the design of in other languages.

Screenshot 2019-06-13 at 14 58 56

Screenshot 2019-06-13 at 14 58 18

File: src/linkcommand.js [hunk 4]
```diff
@@ -25,6 +26,20 @@ export default class LinkCommand extends Command {
...
+
+		/**
+		 * Keeps collection of {@link module:link/utils~ManualDecorator}
+		 * recognized from {@link module:link/link~LinkConfig#decorators}.
+		 * You can consider it as a model of states for custom attributes added to links.
+		 *
+		 * @readonly
+		 * @type {module:utils/collection~Collection}
+		 */
+		this.customAttributes = new Collection();
```
[the remark this answers is not in the document]
Comment (msamsel) [reply src/linkcommand.js]: I rename almost everywhere custom attributes to manual decorator to preserve consistency.

File: src/link.js [hunk 5]
```diff
@@ -34,3 +34,150 @@ export default class Link extends Plugin {
...
+ *
+ * Link decorators are defined as an array of objects:
+ *
+ *		const linkConfig = {
+ *			decorators: [
+ *				{
+ *					mode: 'automatic',
+ *					callback: url => url.startsWith( 'http://' ),
+ *					attributes: {
+ *						target: '_blank',
```
Comment (oleq) [inline src/link.js]: How to configure a value-less HTML attribute? 

For instance, if the [`download`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#Attributes) has a value, it is reflected in the name of the downloaded file. So we cannot use `download: true` here safely because it will download the file named "true".

File: src/link.js [hunk 6]
```diff
@@ -34,3 +34,150 @@ export default class Link extends Plugin {
...
+ * Describes a manual link {@link module:link/link~LinkConfig#decorators decorator}. This kind of a decorator is represented in
+ * the link feature's {@link module:link/linkui user interface} as a switch the user can use to control the presence
+ * of a pre–defined set of attributes.
+ *
+ * For instance, to allow users to manually control the presence of the `target="_blank"` and
+ * `rel="noopener noreferrer"` attributes on specific links, the decorator could look as follows:
+ *
+ *		{
+ *			mode: 'manual',
+ *			label: 'Open link in new window',
```
Comment (oleq) [inline src/link.js]: We need to document how to translate those labels. In the [heading configuration](https://ckeditor5.github.io/docs/nightly/ckeditor5/latest/api/module_heading_heading-HeadingConfig.html#member-options) you can specify some standard strings which are [automatically translated](https://github.com/ckeditor/ckeditor5-heading/blob/master/src/utils.js#L17). Same for [alignment](https://github.com/ckeditor/ckeditor5-alignment/blob/master/src/alignmentui.js#L51), font color, etc..

I think the same should happen here at least for `'Open link in new tab'` and developers must know about it.

File: src/link.js [hunk 7]
```diff
@@ -34,3 +34,150 @@ export default class Link extends Plugin {
...
+ * the `config.link.targetDecorator` option should remain `undefined` or `false` to not interfere with the manual decorator.
+ *
+ * **Note**: It is possible to add other {@link module:link/link~LinkDecoratorAutomaticOption automatic}
+ * or {@link module:link/link~LinkDecoratorManualOption manual} link decorators when this option is active.
+ *
+ * More information about decorators can be found in the {@link module:link/link~LinkConfig#decorators decorators configuration}
+ * reference.
+ *
+ * @default false
+ * @member {Boolean} module:link/link~LinkConfig#targetDecorator
```
Comment (oleq) [inline src/link.js]: I think this config name does not say much as to what it actually does.

Something `config.link.addTargetToExternalLinks` would be more specific.

File: src/linkediting.js [hunk 8]
```diff
@@ -61,13 +77,100 @@ export default class LinkEditing extends Plugin {
...
+	 * {@link module:link/linkcommand~LinkCommand#manualDecorators} collection (a model for manual decorators state).
+	 *
+	 * Also registers an {@link module:engine/conversion/downcasthelpers~DowncastHelpers#attributeToElement attributeToElement}
+	 * converter for each manual decorator and extends the {@link module:engine/model/schema~Schema model's schema}
+	 * with adequate model attributes.
+	 *
+	 * @private
+	 * @param {Array.<module:link/link~LinkDecoratorManualOption>} manualDecoratorDefinitions
+	 */
+	_enableManualDecorators( manualDecoratorDefinitions ) {
```
Comment (oleq) [inline src/linkediting.js]: The function `_enableManualDecorators` accepts `manualDecoratorDefinitions` which are `LinkDecoratorManualOption`. It's counterintuitive.

We should either rename the `LinkDecoratorManualOption` type to `LinkDecoratorManualDefinition` (which makes sense to me) or use `manualDecoratorOptions` as an argument.

File: src/ui/linkformview.js [hunk 9]
```diff
@@ -106,24 +144,23 @@ export default class LinkFormView extends View {
 			}
 		} );
 
+		const classList = [ 'ck', 'ck-link-form' ];
+
+		if ( this.manualDecorators.length ) {
+			classList.push( 'ck-link-form_decorators' );
```
Comment (oleq) [inline src/ui/linkformview.js]: Maybe let's make it more generic like `ck-link-form_layout-vertical`? There could be other things in the future that will toggle the layout to vertical while not being decorators.

If not, at least change it to `ck-link-form_with-decorators` please.

File: src/ui/linkformview.js [hunk 10]
```diff
@@ -77,6 +84,37 @@ export default class LinkFormView extends View {
...
+
+		/**
+		 * A collection of {@link module:ui/button/switchbuttonview~SwitchButtonView},
+		 * which corresponds to {@link #manualDecorators manual decorators} configured in the editor.
+		 * Populated by {@link #_createManualDecoratorsUIView}.
+		 *
+		 * @readonly
+		 * @type {module:ui/viewcollection~ViewCollection}
+		 */
+		this.manualDecoratorsUIView = this._createManualDecoratorsUIView();
```
Comment (oleq) [inline src/ui/linkformview.js]: It's not a view. It's a view collection. This should be clear in the name of the property.

Also: a public property should not link to private method `_createManualDecoratorsUIView`. I think this property does not need to be public in the first place.

Also: I see that this is a collection of switches so this also must be clear. Something like `manualDecoratorSwitches` would fit better.

File: src/ui/linkformview.js [hunk 11]
```diff
@@ -210,6 +248,84 @@ export default class LinkFormView extends View {
...
 	}
+
+	/**
+	 * Populates {@link module:ui/viewcollection~ViewCollection} of {@link module:ui/button/switchbuttonview~SwitchButtonView}
+	 * made based on {@link #manualDecorators}
+	 *
+	 * @private
+	 * @returns {module:ui/viewcollection~ViewCollection} of Switch Buttons.
+	 */
+	_createManualDecoratorsUIView() {
```
Comment (oleq) [inline src/ui/linkformview.js]: It does not create a view. It creates a collection (of switches), which should be clear in the name of the method.

File: src/ui/linkformview.js [hunk 12]
```diff
@@ -210,6 +248,84 @@ export default class LinkFormView extends View {
...
+		const switches = this.createCollection();
+
+		switches.bindTo( this.manualDecorators ).using( item => {
```
Comment (oleq) [inline src/ui/linkformview.js]: Does the `this.manualDecorators` collection change over time? Are new decorators coming and some other leaving it? If not, why bind `switches` to it?

File: src/ui/linkformview.js (hunk 2 above)
Comment (oleq) [reply src/ui/linkformview.js]: Why is this property public BTW? It's used internally only.

File: src/linkcommand.js [hunk 13]
```diff
@@ -49,12 +70,34 @@ export default class LinkCommand extends Command {
...
+	 *
+	 * To learn more about link decorators, check out the {@link module:link/link~LinkConfig#decorators `config.link.decorators`}
+	 * documentation.
+	 *
 	 * @fires execute
 	 * @param {String} href Link destination.
+	 * @param {Object} [manualDecoratorIds={}] The information about manual decorator attributes to be applied or removed upon execution.
 	 */
-	execute( href ) {
+	execute( href, manualDecoratorIds = {} ) {
```
Comment (oleq) [inline src/linkcommand.js]: How to add new decorator attributes and how to remove them (what to pass in `manualDecoratorIds` in both cases)?  This needs an example here.

File: src/link.js [hunk 14]
```diff
@@ -34,3 +34,150 @@ export default class Link extends Plugin {
...
+ * two types of link decorators:
+ *
+ * * {@link module:link/link~LinkDecoratorAutomaticOption automatic} – they match links against pre–defined rules and
+ * manage their attributes based on the results,
+ * * {@link module:link/link~LinkDecoratorManualOption manual} – they allow users to control link attributes individually
+ *  using the editor UI.
+ *
+ * Link decorators are defined as an array of objects:
+ *
+ *		const linkConfig = {
```
Comment (oleq) [inline src/link.js]: ## Idea to discuss

Reading this piece of code an idea popped into my mind that it would be better if the decorators were named. 
I know this is shorter

```js
const linkConfig = {
	decorators: [
		{
			mode: 'automatic',
			callback: url => url.startsWith( 'http://' ),
			attributes: {
				target: '_blank',
				rel: 'noopener noreferrer'
			}
		},
		{
			mode: 'manual',
			label: 'Downloadable',
			attributes: {
				download: 'file.png',
			}
		},
		// ...
	]
}
```

but this would give us clearer attribute names in the model we can use in the command:

```js
const linkConfig = {
	decorators: [
		addTargetToExternals: {
			mode: 'automatic',
			callback: url => url.startsWith( 'http://' ),
			attributes: {
				target: '_blank',
				rel: 'noopener noreferrer'
			}
		},
		downloadableDecorator: {
			mode: 'manual',
			label: 'Downloadable',
			attributes: {
				download: 'file.png',
			}
		},
		// ...
	]
}
```

ATM there's no easy way for developers to figure out which decorator is responsible for what because they are numbered

image

File: src/link.js (hunk 6 above)
Comment (msamsel) [reply src/link.js]: @oleq If I understand correctly `t()` method accepts **only** strings, at least this is how I understand how it works after talk with PK.
So this case is completely different case than Heading, Font, Alignment, etc. We don't have strings for those labels, because those are defined by the developer, so we cannot provide any translations for them.
Originally there was translations for the labels, but I removed them when I noticed this situation and errors during snippet building.
https://github.com/ckeditor/ckeditor5-link/pull/222/commits/83cf302239663a7384529f79d75f9fc76d5c2fd6

If I understand this incorrectly, please give me more info how to approach to translation for unknown strings.

File: src/ui/linkformview.js (hunk 9 above)
Comment (msamsel) [reply src/ui/linkformview.js]: Name of class in `theme-lark` is also corrected.

File: src/ui/linkformview.js (hunk 12 above)
Comment (msamsel) [reply src/ui/linkformview.js]: no, those should be read from config only during editor initialization.

File: src/ui/linkformview.js (hunk 2 above)
Comment (msamsel) [reply src/ui/linkformview.js]: I'd went even further and remove it completely.

File: src/link.js (hunk 6 above)
Comment (oleq) [reply src/link.js]: Generally speaking, you're right. Developers define those strings in the editor configuration, when `t()` is unavailable. 

But just like in the heading feature, we can hardcode certain translations for the most popular options and tell developers that they can use those if they want them to be translated.

In the heading feature, it's developers who define [`config.heading.options`](https://ckeditor5.github.io/docs/nightly/ckeditor5/latest/api/module_heading_heading-HeadingConfig.html#member-options). They can customize them to their liking but as long as they use `'Paragraph'` or `'Heading 3'` in `Title`, those strings will translate to other languages like French or German.

We can do the same for the `'Open in a new tab'` toggle switch option. If enabled as a manual decorator, regardless of the decorator configuration (e.g. with `rel` or without), the thing will always translate. Same with `'Downloadable'`, which is another label people are gonna use.

File: src/link.js (hunk 14 above)
Comment (msamsel) [reply src/link.js]: I like the idea of having names for each decorator. 👍

Comment (msamsel) [reply src/link.js]: Currently I remain it untouched, as such modification will produce lot of docs changes. Let's w8 for more comments if we go into this way with configuration options.

File: src/link.js (hunk 5 above)
Comment (msamsel) [reply src/link.js]: This case as we discuss is rather engine issue, and should be extracted as separate. If `engine.view.elemendefinition` start support attributes without value, decorators will be able to use it as well.

**Comment by oleq** [comment]
This sort of a bug 😛

Kapture 2019-06-19 at 14 19 42

File: src/link.js (hunk 5 above)
Comment (msamsel) [reply src/link.js]: Extracted to:
https://github.com/ckeditor/ckeditor5-engine/issues/1750

File: src/link.js (hunk 14 above)
Comment (mlewand) [reply src/link.js]: Named decorators sound like a good idea. Although I'm a bit worried about a possibility of attribute collision with existing model attr. Say someone names the decorator "bold", "linkHref" or something alike.

Prefixing them with `linkDecorator` should do the job.

**Comment by mlewand** [comment]
The new design looks pretty nice. I'm only concerned about 

> Closing confirmation

That I mentioned in https://github.com/ckeditor/ckeditor5-link/pull/222#issuecomment-493942010 - but if it worries us that's totally something that we should push to a follow-up issue.

File: src/utils.js [hunk 15]
```diff
@@ -59,3 +59,35 @@ function isSafeUrl( url ) {
...
+ * @param {module:core/editor/editor~Editor} editor An editor instance
+ * @returns {Array.<module:link/link~LinkDecoratorAutomaticDefinition|module:link/link~LinkDecoratorManualDefinition>}
+ */
+export function getLocalizedDecorators( editor ) {
+	const t = editor.t;
+	const decorators = editor.config.get( 'link.decorators' );
+
+	if ( decorators ) {
+		const localizedDecoratorsLabels = {
+			'Open in a new tab': t( 'Open in a new tab' ),
```
Comment (oleq) [inline src/utils.js]: Missing entries in `lang/contexts.json`.

**Comment by msamsel** [comment]
> The new design looks pretty nice. I'm only concerned about
> 
> > Closing confirmation
> 
> That I mentioned in [#222 (comment)](https://github.com/ckeditor/ckeditor5-link/pull/222#issuecomment-493942010) - but if it worries us that's totally something that we should push to a follow-up issue.

I feel like introducing such feature might take some time. If I understand correctly we should implement some sort of feedback for the user that current state is "dirty" and confirm if he really want to close edition.
I think that best option could be implementing new feature to notification which utilize `window.confirm` in a similar way as it is done now for [`warnings`](https://ckeditor.com/docs/ckeditor5/latest/api/module_ui_notification_notification-Notification.html#function-showWarning). That's why maybe we could implement it as further improvements ticket.

@oleq, @mlewand WDYT?

**Comment by msamsel** [comment]
> You fixed a bug. Would be good to have a test for that in order to don't fix the same problem in the future.

Thx for feedback. I added unit tests for this change now.

File: tests/manual/linkdecorator.js [hunk 16]
```diff
@@ -21,30 +21,30 @@ ClassicEditor
 		plugins: [ Link, Typing, Paragraph, Clipboard, Undo, Enter ],
 		toolbar: [ 'link', 'undo', 'redo' ],
 		link: {
-			decorators: [
-				{
+			decorators: {
+				IxExternal: {
```
Comment (oleq) [inline tests/manual/linkdecorator.js]: `is`. Also most likely `isExternal` because we use camel case for props in the project.

**Comment by oleq** [comment]
But... you didn't change the configuration documentation, did you @msamsel?

**Comment by msamsel** [comment]
> But... you didn't change the configuration documentation, did you @msamsel?

I'm still in progress. Right now I just implement new config and correct tests and now switch to docs.

File: theme/linkform.css [hunk 17]
```diff
@@ -26,3 +24,27 @@
 		}
 	}
 }
+
+/*
+ * Style link form differently when manual decorators are available.
+ * See: https://github.com/ckeditor/ckeditor5-link/issues/186.
+ */
+.ck.ck-link-form_layout-vertical {
+	flex-wrap: wrap;
```
Comment (oleq) [inline theme/linkform.css]: How is this supposed to work if `ckeditor5-link/theme/linkform.css` defines this class with `display: block` (not flex?) https://github.com/ckeditor/ckeditor5-theme-lark/pull/233/files#diff-e9ba6bcf18c57cfa9c06993af81ac820R54

**Comment by msamsel** [comment]
@oleq I see that @dkonopka made requested changes.
